src/StreamJsonRpc/JsonMessageFormatter.cs

```diff
@@ -518,5 +649,106 @@ internal JsonArrayPool(ArrayPool<T> arrayPool)
 
             public void Return(T[] array) => this.arrayPool.Return(array);
         }
+
+        private class JsonProgressClientConverter : JsonConverter
+        {
+            private readonly JsonMessageFormatter formatter;
+
+            public JsonProgressClientConverter(JsonMessageFormatter formatter)
+            {
+                this.formatter = formatter ?? throw new ArgumentNullException(nameof(formatter));
+            }
+
+            public override bool CanConvert(Type objectType) => MessageFormatterHelper.FindIProgressOfT(objectType) != null;
+
+            public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+            {
+                throw new NotSupportedException();
+            }
+
+            public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+            {
+                // if the requestId is empty it means the Progress object comes from a response or a notification
+                if (this.formatter.requestId == null)
+                {
+                    throw new NotSupportedException("IProgress<T> objects should not be part of any response or notification.");
+                }
+
+                lock (this.formatter.progressMap)
+                {
+                    long progressId = this.formatter.nextProgressId++;
+                    this.formatter.requestProgressMap.Add(this.formatter.requestId.Value, progressId);
+
+                    this.formatter.progressMap.Add(progressId, new ProgressParamInformation(value));
+
+                    writer.WriteValue(progressId);
+                }
+            }
+        }
+
+        private class ProgressParamInformation
+        {
+            public Type iProgressOfTType;
+            public Type valueType;
+            public MethodInfo reportMethod;
+            public object progressObject;
+
+            public ProgressParamInformation(object progressObject)
+            {
+                this.iProgressOfTType = MessageFormatterHelper.FindIProgressOfT(progressObject.GetType());
+                this.valueType = this.iProgressOfTType.GenericTypeArguments[0];
```

Comment:
>GenericTypeArguments [](start = 55, length = 20)

Can this run into index out of bounds errors?

Reply:
No, because the method we're calling only returns `IProgress<T>` instances, so we're guaranteed exactly one member of this array.

---
In reply to: [304595863](https://github.com/milopezc/vs-streamjsonrpc/pull/1#discussion_r304595863) [](ancestors = 304595863)